### PR TITLE
Fix listener completion in IndexShard.scheduledRefresh

### PR DESCRIPTION
**Pull request by tlrx**
This method was recently changed and after looking at it I think the listener may never be completed (if verifyNotClosed or getEngine throws) or double-completed with false or pass to maybeRefresh and then again completed with false.

### CODE REVIEW

**Review by DaveCTurner — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -3747,28 +3747,34 @@ ReplicationTracker getReplicationTracker() {
...
-                // lets skip this refresh since we are search idle and
-                // don't necessarily need to refresh. the next searcher access will register a refreshListener and that will
-                // cause the next schedule to refresh.
-                engine.maybePruneDeletes(); // try to prune the deletes in the engine if we accumulated some
-                setRefreshPending(engine);
-                ActionListener.completeWith(listener, () -> false);
-            } else {
-                logger.trace("refresh with source [schedule]");
-                engine.maybeRefresh("schedule", listener.map(Engine.RefreshResult::refreshed));
+        try {
```
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:3750]: nit You could replace the try/catch with ActionListener.run().

**Review by kingherc — COMMENTED** [review]
@tlrx I see the issues with the listeners, but I am not sure how it fixes the CI failure. The code that failed in the test is

```
assertTrue(future5.actionGet());
```

which asserts that the listener should have returned true. The code changes in this PR, however, are for failures and exceptions that would return false. I am not seeing how the PR helps returns true?

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 1 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I pushed 50940c26fd7d6fd9190a38994aec7700a0ab99b9

**Comment by tlrx** [comment]
@kingherc Sorry, I wasn't clear. This change does not fix #96920, [my comment](https://github.com/elastic/elasticsearch/issues/96920#issuecomment-1602804939) there was just to raise awareness that I spotted issues with the listeners.

**Review by kingherc — APPROVED** [review]
Ah thanks for explaining :)

**Comment by tlrx** [comment]
Thanks David and Iraklis!

**Pull request merged**